### ISSUE TITLE
Linux - Save training data seems to export empty cameras bin file

### ISSUE THREAD

**Issue by nosmose** [david-rhodes/GSOPs#4]
Hi !

Wonderful piece of code !

Trying to export the training data from Linux to COLMAP, and to postshot.
I've got the points3d exported nicely, got cameras.bin and images.bin files created too.
But loading into COLMAP it seems like the cameras.bin file is empty (only 64bytes on disk)

Trying to debug but nothing visible in the logs, only the saved log for the bin files.

Any ideas of what could be the issue?

**Comment by david-rhodes** [comment]
Hello @nosmose, 

I have not tested GSOPs with Linux, so it's likely there are bugs, especially as it relates to system-specific I/O. 

To help me debug this problem, do you get any errors when saving training data? If you "allow editing of contents" on the `generate_training_data` SOP and dive inside, do you see any nodes with warnings or errors afterwards?

Did you provide a camera operator path in the camera settings tab?

If you're comfortable, you could put a `print(cameras)` line in the HDA (type properties-->scripts tab) at the bottom of the `write_cameras_bin` method. This might give us some additional clues. 

Also, could you please upload your bad .bin file here, please?

**Comment by nosmose** [comment]
Hi David !
Thanks for your fast answer!

Totally understood it's not tested on linux, but so far it has been pretty flawless apart from this issue.
So looking into the training_data node, the only node erroring is SFM_MODEL_ALIGNMENT_EXPORT, related to the COLMAP path don't matching. 
But I don't think that's related isnt it?

In the camera settings there is indeed a camera op path, pointing to a proper generated camera.

Thanks for the type properties > script tab tips, I had trouble finding where the scripts were actually located. I'll try to debug using some prints.

I'm attaching here the cameras.bin and the images.bin, not sure yet if that second one works too to be honest

Thanks a lot !
[gsops_bin.zip](https://github.com/david-rhodes/GSOPs/files/14594221/gsops_bin.zip)

**Comment by david-rhodes** [comment]
Ok, so at first glance the `cameras.bin` is fine: 
`{1: Camera(id=1, model='PINHOLE', width=1280, height=720, params=array([1545.09504749, 1545.095
04749,  640.        ,  360.        ]))}`

I tried to read the images.bin with `colmap.read_write_model`, but it fails with the following:

`Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "<console>", line 15, in read_images_bin
  File "/Users/david/Documents/GitHub/GSOPs/scripts/python/colmap/read_write_model.py", line 25
7, in read_images_binary
    binary_image_properties = read_next_bytes(
  File "/Users/david/Documents/GitHub/GSOPs/scripts/python/colmap/read_write_model.py", line 87
, in read_next_bytes
    return struct.unpack(endian_character + format_char_sequence, data)
struct.error: unpack requires a buffer of 64 bytes`

Does this happen with the example scenes?

**Comment by nosmose** [comment]
Indeed after printing out the data from the cameras.bin and images.bin it looks accurate and complete.

I try to validate the data importing the model into COLMAP on my windows machine.
But looking at the log looks like it found 0 cameras, 0 matches and 0 images.

I tried using the example scene, it manages to load cameras into COLMAP, but still getting the same log output 0 cameras, 0 matches and 0 images.

The culprit might be not on the houdini gsops side after all.

**Comment by david-rhodes** [comment]
Ok, this is quite helpful information--thank you!

How are you opening the model in COLMAP? I've only had success with File-->Import Model (pointing to the `sparse/0` directory). With this approach, I was able to load your model (639 image/camera pairs?), though I had to provide my own points3D.bin (which is a feature just recently added to the `develop` branch).

![image](https://github.com/david-rhodes/GSOPs/assets/23202691/ca2b1c79-4a3b-4a27-8e30-fb28cb5ec725)

It's also important to understand that GSOPs does not generate a complete images.bin file. For example, I don't save any "tracked points" data, as this is not required for training 3DGS models. However, that data is required to preview the points3D.bin (by default). To show the points with the cameras, follow this guide:

![image](https://github.com/david-rhodes/GSOPs/assets/23202691/7e65a474-9405-4af3-a2b0-c124a0deff7d)

To summarize, your custom data may be valid after all. Just be sure to export a points3D.bin as well as cameras.bin and images.bin. Also, your cameras seem like they are very far apart (I had to zoom the scene way out and increase the camera icon size to even see them). This could cause potential issues? It might be worth validating against the example scene's exported data.

**Comment by nosmose** [comment]
wow thanks a lot for your help
I was using your latest branch with a points3d.bin already, but seeting the point min. track length made it visible
also the scale of the scene is quite huge (it's a full environment) hence the need to scale way up the camera.
I am able to visualize it fully now.

So yeah the custom data is totally valid, it was more me poorly using COLMAP.

thanks a bunch

**Comment by david-rhodes** [comment]
You're welcome! 

For the best experience, I recommend training your data with [Postshot](https://www.jawset.com/). Simply drag and drop your images and sparse folders (at the same time) into the application and go from there. 😄

**Comment by david-rhodes** [comment]
What do you mean by images? When I drop my COLMAP data into Postshot, I only see camera frustum representations and the sparse point cloud (no actual images). 

Does it train? The only thing I can think of is that it's a scene scale issue. If possible, you could try to "normalize" your scene before  using generate training data (this might speed up your camera pose generation, too).

**Comment by nosmose** [comment]
you're correct I don't see the images but the log show them as loaded.
It does train but the result shows no radiance field.
The main difference is that with your example scene, when I go in postshot, I can see the cameras position/orientation infos changing per frame. Whereas with my scene I only get one camera at the center of world, no values in position/orientation, like if the cameras werent loading.

I will give it a try with a simplier scene maybe, something with a more simple scale to see if that's part of the issue.
